torch/futures/__init__.py

```diff
@@ -81,6 +81,39 @@ def then(self, callback):  # type: (Callable[[Future[T]], S]) -> Future[S]
         """
         return cast(Future[S], super().then(callback))
 
+    # Have to use string annotations because  PEP-0563 is not available in 3.6
+    def add_done_callback(self, callback):  # type: (Callable[[Future[T]], None]) -> None
+        r"""
+        Append the given callback function to this ``Future``, which will be run
+        when the ``Future`` is completed.  Multiple callbacks can be added to
+        the same ``Future``, and will be invoked in the same order as they were
+        added. The callback must take one argument, which is the reference to
+        this ``Future``. The callback function can use the ``Future.wait()`` API
+        to get the value.
+
+        We recommend that you use the ``then`` API as it provides a way to synchronize
```

Comment:
Added some clarifications to the doc after talking with Greg. I don't really see a use case where `add_done_callback` gives you someone that `then` does not, so I added recommendations to the doc. I also clarified what actually happens when you interleave the two calls, since that's probably behaviors that users would want to know about (even though they probably shouldn't do that).